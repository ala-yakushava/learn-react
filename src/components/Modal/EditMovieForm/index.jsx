import React from 'react';
import PropTypes from 'prop-types';
import { useFormik } from 'formik';

import './style.scss';
import Heading from '../../Heading';
import TextInput from '../../TextInput';
import ModalFooter from '../ModalFooter';
import Button from '../../Button';
import { formField } from '../../../utils/data';
import MovieFormSchema from '../../../utils/validate';

const EditMovieForm = ({ movie, onSubmit }) => {
  const defaultState = {
    title: movie.title,
    release_date: movie.release_date,
    poster_path: movie.poster_path,
    genres: movie.genres ? movie.genres.join(', ') : '',
    overview: movie.overview,
    runtime: String(movie.runtime),
  };

  const formik = useFormik({
    initialValues: defaultState,
    validationSchema: MovieFormSchema,
    onSubmit: (values) => {
      const data = {
        ...values,
        id: movie.id,
        runtime: Number(values.runtime),
        genres: values.genres.split(', '),
      };
      onSubmit(data);
    },
  });

  const {
    values, isSubmitting, dirty, isValid, errors, handleSubmit, resetForm, handleChange,
  } = formik;

  return (
    <form
      className="EditMovieForm"
      onSubmit={handleSubmit}
      onReset={resetForm}
      data-testid="edit-form"
    >
      <Heading>Edit Movie</Heading>
      <TextInput
        value={String(movie.id)}
        readOnly
        label="Movie ID"
      />
      {formField.map(({
        id, name, type, label, placeholder,
      }) => (
        <TextInput
          key={id}
          name={name}
          type={type}
          label={label}
          placeholder={placeholder}
          value={values[name]}
          required
          onChange={handleChange}
          errors={errors[name]}
        />
      ))}
      <ModalFooter>
        <Button type="reset" mode="secondary">Reset</Button>
        <Button type="submit" mode="primary" disabled={isSubmitting || !dirty || !isValid}>Save</Button>
      </ModalFooter>
    </form>
  );
};

EditMovieForm.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    genres: PropTypes.arrayOf(PropTypes.string),
    release_date: PropTypes.string,
    poster_path: PropTypes.string,
    overview: PropTypes.string,
    runtime: PropTypes.number,
  }),
  onSubmit: PropTypes.func,
};

EditMovieForm.defaultProps = {
  movie: {},
  onSubmit: () => {},
};

export default EditMovieForm;
