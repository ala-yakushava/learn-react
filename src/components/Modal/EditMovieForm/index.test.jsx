import React from 'react';
import renderer from 'react-test-renderer';

import EditMovieForm from './index';

describe('Components - EditMovieForm', () => {
  test('should match snapshot', () => {
    const tree = renderer
      .create(<EditMovieForm />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
