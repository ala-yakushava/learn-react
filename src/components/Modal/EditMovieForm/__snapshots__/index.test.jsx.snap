// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - EditMovieForm should match snapshot 1`] = `
<form
  className="EditMovieForm"
  data-testid="edit-form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="Heading"
  >
    Edit Movie
  </div>
  <div
    className="TextInput"
  >
    <label
      className="TextInput_label"
      htmlFor=""
    >
      Movie ID
    </label>
    <input
      className="TextInput_field"
      id=""
      name=""
      onChange={[Function]}
      placeholder={null}
      readOnly={true}
      required={false}
      type="text"
      value="undefined"
    />
  </div>
  <div
    className="TextInput"
  >
    <label
      className="TextInput_label"
      htmlFor="title"
    >
      Title
    </label>
    <input
      className="TextInput_field"
      id="title"
      name="title"
      onChange={[Function]}
      placeholder="Title here"
      readOnly={false}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="TextInput"
  >
    <label
      className="TextInput_label"
      htmlFor="release_date"
    >
      Release date
    </label>
    <input
      className="TextInput_field"
      id="release_date"
      name="release_date"
      onChange={[Function]}
      placeholder="Select date"
      readOnly={false}
      required={true}
      type="date"
      value=""
    />
  </div>
  <div
    className="TextInput"
  >
    <label
      className="TextInput_label"
      htmlFor="poster_path"
    >
      Movie URL
    </label>
    <input
      className="TextInput_field"
      id="poster_path"
      name="poster_path"
      onChange={[Function]}
      placeholder="Movie URL here"
      readOnly={false}
      required={true}
      type="url"
      value=""
    />
  </div>
  <div
    className="TextInput"
  >
    <label
      className="TextInput_label"
      htmlFor="genres"
    >
      Genre
    </label>
    <input
      className="TextInput_field"
      id="genres"
      name="genres"
      onChange={[Function]}
      placeholder="Select genre"
      readOnly={false}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="TextInput"
  >
    <label
      className="TextInput_label"
      htmlFor="overview"
    >
      Overview
    </label>
    <input
      className="TextInput_field"
      id="overview"
      name="overview"
      onChange={[Function]}
      placeholder="Overview text goes here"
      readOnly={false}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="TextInput"
  >
    <label
      className="TextInput_label"
      htmlFor="runtime"
    >
      Runtime
    </label>
    <input
      className="TextInput_field"
      id="runtime"
      name="runtime"
      onChange={[Function]}
      placeholder="Runtime text goes here"
      readOnly={false}
      required={true}
      type="text"
      value="undefined"
    />
  </div>
  <div
    className="ModalFooter"
  >
    <button
      className="Button Button--secondary"
      disabled={false}
      onClick={[Function]}
      type="reset"
    >
      <span
        className="Button_text"
      >
        Reset
      </span>
    </button>
    <button
      className="Button Button--primary"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      <span
        className="Button_text"
      >
        Save
      </span>
    </button>
  </div>
</form>
`;
